src/lib/storage/yaml_writer.rb

```diff
@@ -0,0 +1,141 @@
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "yaml"
+require "storage"
+require "storage/disk_size.rb"
+require "storage/enum_mappings.rb"
+
+module Yast
+  module Storage
+    #
+    # Class to write storage device trees to YAML files.
+    #
+    class YamlWriter
+      include Yast::Logger
+      include EnumMappings
+
+      def initialize
+        # Cache some frequently needed values: We need the inverse mapping from
+        # EnumMappings, i.e. from the C++ enum to string.
+        @partition_table_types = PARTITION_TABLE_TYPES.invert
+        @partition_types       = PARTITION_TYPES.invert
+        @partition_ids         = PARTITION_IDS.invert
+        @file_system_types     = FILE_SYSTEM_TYPES.invert
+      end
+
+      # Write all devices from the specified device graph to a YAML file.
+      #
+      # @param devicegraph [::Storage::devicegraph]
+      # @param yaml_file_name [String]
+      #
+      def write(devicegraph, yaml_file_name)
+        device_tree = yaml_device_tree(devicegraph)
+        File.open(yaml_file_name, "w") { |file| file.write(device_tree.to_yaml) }
+      end
+
+      # Convert all devices from the specified device graph to YAML data
+      # structures, i.e. nested arrays and hashes. The toplevel item will
+      # always be an array.
+      #
+      # @param devicegraph [::Storage::devicegraph]
+      # @return yaml [Array<Hash>]
+      #
+      def yaml_device_tree(devicegraph)
+        devicegraph.all_disks.to_a.inject([]) { |yaml, disk| yaml << yaml_disk(disk) }
+      end
+
+      private
+
+      # Return the YAML counterpart of a ::Storage::Disk.
+      #
+      # @param  disk [::Storage::Disk]
+      # @return yaml [Hash]
+      #
+      def yaml_disk(disk)
+        content = {}
+        content["size"] = DiskSize.new(disk.size_k).to_s
+        content["name"] = disk.name
+        begin
+          ptable = disk.partition_table # this will raise an excepton if no partition table
+          content["partition_table"] = @partition_table_types[ptable.type]
+          cyl_size = DiskSize.new(disk.geometry.cylinder_size / 1024)
+          first_free_cyl = 0
+          partitions = []
+
+          ptable.partitions.each do |partition|
+            gap = partition.region.start - first_free_cyl
+            if gap > 0
+              partitions << yaml_free_slot(cyl_size*gap)
+            end
+
+            partitions << yaml_partition(partition)
+            first_free_cyl = partition.region.end + 1 unless partition.type == ::Storage::EXTENDED
+          end
+          content["partitions"] = partitions unless partitions.empty?
+        rescue RuntimeError => ex # FIXME: rescue ::Storage::Exception when SWIG bindings are fixed
```

Comment:
We are starting to have this in too many places. Do we have a concrete plan to fix those SWIG bindings? We should push for it to be done as soon as possible.